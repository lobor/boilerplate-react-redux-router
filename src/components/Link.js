import React from 'react'
// import { router, navigate } from './../store/'
import { routeStore, navigate } from 'store/router'
import styled from 'styled-components';

const A = styled.a``;

export default class Link extends React.Component {
	state = {
		active: false
	}

	static defaultProps = {
		selected: false,
		className: '',
		regexp: false
	}

	componentWillMount() {
		var reg;
		var state = routeStore.getState();
		if (this.props.regexp) {
			reg = new RegExp(this.props.href, 'g');
		}

		if (state && state.path) {
			this.setState({ active: state.path.path === this.props.href || this.props.selected || reg && state.path.path.match(reg) })
			this.unsubscribe = routeStore.subscribe(() => {
				this.setState({ active: state.path.path === this.props.href || reg && state.path.path.match(reg) })
			})
		}
	}

	componentWillUnmount() {
		this.unsubscribe();
	}

	clickItem = (e) => {

		if (e) {
			e.preventDefault();
			e.stopPropagation();
		}

		if (this.props.onClick) {
			this.props.onClick(e);
		}

		if (e && e.target.href) {
			routeStore.dispatch(navigate(e.target.href))
		}
	}

	render() {
		var { active } = this.state
		var {
			children,
			href,
			className,
			regexp,
			onClick,
			...args
		} = this.props;

		className += active ? ' selected' : ''

		return (<A href={href} onClick={this.clickItem} {...args} className={className}>{children}</A>)
	}
}
