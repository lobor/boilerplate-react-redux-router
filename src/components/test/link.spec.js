import React from 'react'
import styled from 'styled-components'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme';

import Link from './../Link'

describe('Test component Link', () => {
	test('render link', () => {
		var component = renderer.create(<Link />);
	  const tree = component.toJSON()
	  expect(tree).toMatchSnapshot()
	})

	test('render link selected', () => {
		var component = renderer.create(<Link href="/" selected regexp />);
	  const tree = component.toJSON()
	  expect(tree).toMatchSnapshot()
	})

	test('render link click', () => {
		var component = renderer.create(<Link href="/" selected regexp />);
		const tree = component.toJSON()
		tree.props.onClick();
	  expect(component).toMatchSnapshot()
	})

	test('render link click with props onClick', () => {
		var toto = false;
		var component = renderer.create(<Link href="/" selected regexp onClick={() => { toto = true }} />);
		const tree = component.toJSON()
		tree.props.onClick();
	  expect(toto).toBe(true)
	})

	test('render link click with props onClick and event', () => {
		var toto = false;
		var component = renderer.create(<Link href="/" selected regexp onClick={() => { toto = true }} />);
		const tree = component.toJSON()
		tree.props.onClick({preventDefault: () => {}, stopPropagation: () => {}, target: { href: false }});
	  expect(toto).toBe(true)
	})
});
