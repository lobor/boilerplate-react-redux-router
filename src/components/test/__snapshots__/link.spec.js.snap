// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component Link render link 1`] = `
<a
  className="sc-bdVaJa bDWFJH"
  href={undefined}
  onClick={[Function]}
  selected={false}
/>
`;

exports[`Test component Link render link click 1`] = `
<a
  className="sc-bdVaJa bDWFJH"
  href="/"
  onClick={[Function]}
  selected={true}
/>
`;

exports[`Test component Link render link selected 1`] = `
<a
  className="sc-bdVaJa bDWFJH"
  href="/"
  onClick={[Function]}
  selected={true}
/>
`;
